### DISCUSSION TITLE
Support for composite builds

### DISCUSSION THREAD

**Discussion by everett1992** [npm/feedback#371] [CLI] [answered]
### Problem statement

I'm working with a build system that coordinates builds of multiple repositories.  Users define a directed acyclic graph of repository commits and the build system will build those repositories in topological order. Each build can access output files of the previous build. Users can change a repositories commit in the DAG and the system will rebuild all of it's consumers from the new change. Users can also checkout repositories into a local workspace and rebuild the graph with working tree changes. 

We're struggling to integrate this system with npm. We want to use lockfiles to prevent changing versions installed from the npm registry but modules defined in the DAG should not be locked. Most npm dependency specifiers are immutable or are included in the lockfile as immutable.

We cannot use local path dependencies because the relative path from a module to the build output is unpredictable. `npm link` doesn't work because you need to `npm install` first to create node_modules with registry dependencies and there may not be anything to install from the registry for module in the DAG.

workspaces seem promising, after all our build system is similar to a monorepo, however workspaces cannot be composed, so we could not use a workspace to define our DAG and allow repositories to use the workspaces feature.

[overrides](https://github.com/npm/rfcs/blob/latest/accepted/0036-overrides.md) seem like they may work, tho I'm not sure how they interact with lock files. The rfc defines overrides in package.json while we would like to set overrides dynamically and would prefer to do so in a file that isn't normally included in source control.

### Current solution
This is a total hack. We run a localhost npm registry for each DAG build (for unrelated reasons). That registry allows re-publishing the same package version. Before running `npm install` we remove remove the `integrity` field from all packages that were published to the localhost registry from package-lock.json. We also update the `dependencies` section in each published pacakge in the package-lock to reflect the latest published version.

This has 'worked'  with lockfile v1 but is dubiously correct. It will be harder to adapt the algorithm to v2 and 3 `packages` section. 

### Proposed solutions
The overrides proposal may work, but I don't know how it interacts with package-lock. If overrides supersedes lockfiles we could generate overrides for packages in the DAG. We'd like to specify overrides in a file that isn't checked into source.
```
{
  "overrides": {
    "foo": "file:<root>/outputs/FooRepository/"
   }
}
```

Other options to consider are nested workspaces, automatic `npm link`, a way to exclude packages from package-lock, or a way to specify path dependencies from an alias `file:<prefix>/foo`

**Comment by MylesBorins** [comment, marked as answer]
Hey @everett1992. If you have a specific proposal for the npm CLI the next step would be to open an issue over at https://github.com/npm/rfcs to make a proposal. 

I believe that there may some interest in longer term support for nested workspaces, @darcyclarke may be able to add more.

**Reply by everett1992** [reply]
I wanted to start in discussions because I'm not ready to make a specific proposal. I'm not sure if overrides or workspaces will satisfy our needs. I did some more testing today and saw that `workspaces` are included in lockfile as links resolved to the workspace path. That precluded dynamically setting workspace paths.
